### PR TITLE
Make sure we don't use un-initialized reader when a command has no fields struct.

### LINKED ISSUE #10501: A command with no data struct will cause use of an un-initialized TLV reader
#### Problem
`CommandHandler::ProcessCommandDataElement` does this:
```
    chip::TLV::TLVReader commandDataReader;
...

    err = aCommandElement.GetData(&commandDataReader);
    if (CHIP_END_OF_TLV == err)
    {
        err = CHIP_NO_ERROR;
        ChipLogDetail(DataManagement, "Received command without data for cluster " ChipLogFormatMEI, ChipLogValueMEI(clusterId));
    }
```
But now commandDataReader is not initialized, and we proceed to pass it to `DispatchSingleClusterCommand` which will try to read from it.

 #### Proposed Solution
1. Write a test that exercises this case.  In particular, this should send no data for a command that normally _does_ have fields that get sent.
2. Fix the code to handle this correctly.

@yunhanw-google @mrjerryjohns

### PR DESCRIPTION
**Pull request by bzbarsky-apple**
Fixes https://github.com/project-chip/connectedhomeip/issues/10501

#### Problem
See #10501

#### Change overview
If no fields struct was sent, point payload reader to an empty TLV struct.

#### Testing
Unit tests updated; the tests failed without the code changes.

### CODE REVIEW

**Review by tcarmelveilleux — APPROVED** [review]

**Review by yunhanw-google — APPROVED** [review]

File: src/app/tests/TestCommandInteraction.cpp [hunk 1]
```diff
@@ -224,14 +259,14 @@ class TestExchangeDelegate : public Messaging::ExchangeDelegate
...
-CommandPathParams MakeTestCommandPath(CommandId aCommandId = kTestCommandId)
+CommandPathParams MakeTestCommandPath(CommandId aCommandId = kTestCommandIdWithData)
 {
     return CommandPathParams(kTestEndpointId, 0, kTestClusterId, aCommandId, (chip::app::CommandPathFlags::kEndpointIdValid));
 }
 
 void TestCommandInteraction::GenerateInvokeRequest(nlTestSuite * apSuite, void * apContext, System::PacketBufferHandle & aPayload,
                                                    bool aNeedCommandData, bool aIsTimedRequest, EndpointId aEndpointId,
-                                                   ClusterId aClusterId, CommandId aCommandId)
+                                                   ClusterId aClusterId, Optional<CommandId> aCommandId)
```
Comment (erjiaqing) [inline src/app/tests/TestCommandInteraction.cpp:269]: Would suggest using a extra `aWithData` flag, Optional<CommandId> is hard to understand

File: src/app/CommandHandler.cpp [hunk 2]
```diff
@@ -244,6 +244,20 @@ CHIP_ERROR CommandHandler::SendCommandResponse()
...
+// We use this when the sender did not actually provide a CommandFields struct,
+// to avoid downstream consumers having to worry about cases when there is or is
+// not a struct available.
+//
+// The encoding here is:
+// Control byte: Struct, anonymous tag (since we can't use a context tag at top
+//               level, and consumers should not care about the tag here).
+// End of container.
+constexpr uint8_t sNoFields[] = {
+    0b00010101 /* struct, anonymous tag */, 0b00011000 /* End of container */
```
Comment (mrjerryjohns) [inline src/app/CommandHandler.cpp:257]: Please use the TLV macros instead please. Those avoid the direct coupling to internal TLV encoding specifics:

```
constexpr uint8_t sNoFields[] = {
            CHIP_TLV_STRUCTURE(CHIP_TLV_TAG_ANONYMOUS),
            CHIP_TLV_END_OF_CONTAINER,
}
```

**Review by mrjerryjohns — CHANGES_REQUESTED** [review]

File: src/app/CommandHandler.cpp (hunk 2 above)
Comment (bzbarsky-apple) [reply src/app/CommandHandler.cpp:257]: Did not know about those.  Yes, that makes a lot of sense!

File: src/app/tests/TestCommandInteraction.cpp (hunk 1 above)
Comment (bzbarsky-apple) [reply src/app/tests/TestCommandInteraction.cpp:269]: We can do that, looks like, given the limited number of callers.

**Review by mrjerryjohns — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

**Pull request merged**